kratos/processes/from_json_check_result_process.cpp

```diff
@@ -0,0 +1,866 @@
+                    }
+                }
+            }
+        } else { // Non-historical values
+            for (auto& p_var_double : mpNodalVariableDoubleList) {
+                const auto& r_var_database = r_node_database.GetVariableData(*p_var_double);
+
+                #pragma omp parallel for reduction(+:check_counter)
+                for (int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+                    auto it_node = it_node_begin + i;
+
+                    const double result = it_node->GetValue(*p_var_double);
+                    const double reference = r_var_database.GetValue(i, time);
+                    if (!CheckValues(result, reference)) {
+                        FailMessage(it_node->Id(), "Node", result, reference, p_var_double->Name());
+                        check_counter += 1;
+                    }
+                }
+            }
+            for (auto& p_var_array : mpNodalVariableArrayList) {
+                const auto& r_var_database = r_node_database.GetVariableData(*p_var_array);
+
+                #pragma omp parallel for reduction(+:check_counter)
+                for (int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+                    auto it_node = it_node_begin + i;
+
+                    const auto& r_entity_database = r_var_database.GetEntityData(i);
+                    const array_1d<double, 3>& r_result = it_node->GetValue(*p_var_array);
+                    for (IndexType i_comp = 0; i_comp < 3; ++i_comp) {
+                        const double reference = r_entity_database.GetValue(time, i_comp);
+                        if (!CheckValues(r_result[i_comp], reference)) {
+                            FailMessage(it_node->Id(), "Node", r_result[i_comp], reference, p_var_array->Name());
+                            check_counter += 1;
+                        }
+                    }
+                }
+            }
+            for (auto& p_var_vector : mpNodalVariableVectorList) {
+                const auto& r_var_database = r_node_database.GetVariableData(*p_var_vector);
+
+                #pragma omp parallel for reduction(+:check_counter)
+                for (int i = 0; i < static_cast<int>(r_nodes_array.size()); ++i) {
+                    auto it_node = it_node_begin + i;
+
+                    const auto& r_entity_database = r_var_database.GetEntityData(i);
+                    const Vector& r_result = it_node->GetValue(*p_var_vector);
+                    for (IndexType i_comp = 0; i_comp < r_result.size(); ++i_comp) {
+                        const double reference = r_entity_database.GetValue(time, i_comp);
+                        if (!CheckValues(r_result[i_comp], reference)) {
+                            FailMessage(it_node->Id(), "Node", r_result[i_comp], reference, p_var_vector->Name());
+                            check_counter += 1;
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Correct me if I'm wrong, but as far as I see the code is identical for the historical and non-historical variables (aside of the Get's). If so it is, maybe it is possible to create an intermediate templated method to avoid the code repetition. Something like
```
if (this->Is(HISTORICAL_VALUE)) {
    the_new_method<true>(args);
} else {
    the_new_method<false>(args);
}
```
Then we create an auxiliary Get method which is specialized with the FastGet.. and simple Get accordingly.

Reply:
The problem is the efficeny, but I can do it using template magic

Reply:
If you do it with a template it think it shouldn't affect the runtime performance.